File: sheet_ops_js/__tests__/sheetOps.test.js
import updateSpreadsheet from "../index.js"
import * as fs from 'fs'
import GoogleSheetsApiMock from "./GoogleSheetsApiMock.js"

function simulate(current, target) {
  const mock = new GoogleSheetsApiMock()
  mock.addSpreadsheet("abc", current)
  updateSpreadsheet(mock, "abc", current, target)
  return mock.fetchSpreadsheetState("abc")
}

test("empty", () => {
  const current = { columns: [], rows: [] }
  const target = { columns: [], rows: [] }
  expect(simulate(current, target)).toStrictEqual(target)
})

test("nulls", () => {
  const current = { columns: [null, null, null], rows: [null, null, null] }
  const target = { columns: [null, null, null], rows: [null, null, null] }
  expect(simulate(current, target)).toStrictEqual(target)
})

test("example 1", () => {
  const current = {
    columns: [14, null, null, null, 12, 7, 4, 13, 6, null],
    rows: [15, 14, 13, null, 8, 11, null, 7, 12, 2, null, null, 9]
  }
  const target = {
    columns: [null, null, 11, 14, null, 4, null, 9, 13, 5, 7, 3],
    rows: [9, 12, 3, 13, null, null, 4, 8, 5, null, 6, null, 10]
  }
  expect(simulate(current, target)).toStrictEqual(target)
})

test("example 2", () => {
  const current = {
    columns: [8, 7, 15, null, 11, 13, 2],
    rows: [null, 2, 6, 7, 8, 1, 5]
  }
  const target = {
    columns: [4, 5, null, 2, 6, 1, 8, 7],
    rows: [8, 6, 5, 7, 4, null, 1]
  }
  expect(simulate(current, target)).toStrictEqual(target)
})

test("large sheet", () => {
  const current = JSON.parse(fs.readFileSync('./__tests__/data/large_current.json'))
  const target = JSON.parse(fs.readFileSync('./__tests__/data/large_target.json'))
  const final = simulate(current, target)
  expect(final).toStrictEqual(target)
})

File: sheet_ops_clj/test/sheet_ops_test.clj
(ns sheet-ops-test
  (:require [clojure.data.json :as json]
            [clojure.test :refer [deftest is testing use-fixtures]]
            [google-sheet-api-mock]
            [schema.test]
            [sheet-ops]))

(use-fixtures :once schema.test/validate-schemas)

(defn simulate [current target]
  (let [mock (google-sheet-api-mock/new-mock)
        spreadsheet-id "test"]
    (google-sheet-api-mock/add-spreadsheet! mock spreadsheet-id current)
    (sheet-ops/update-spreadsheet mock spreadsheet-id current target)
    (google-sheet-api-mock/get-spreadsheet mock spreadsheet-id)))

(deftest update-spreadsheet
  (testing "empty"
    (let [current {:columns [], :rows []}
          target {:columns [], :rows []}]
      (is (= target (simulate current target)))))

  (testing "nils"
    (let [current {:columns [nil nil nil], :rows [nil nil]}
          target {:columns [nil nil nil], :rows [nil nil]}]
      (is (= target (simulate current target)))))

  (testing "example 1"
    (let [current {:columns [14, nil, nil, nil, 12, 7, 4, 13, 6, nil]
                   :rows [15, 14, 13, nil, 8, 11, nil, 7, 12, 2, nil, nil, 9]}
          target {:columns [nil, nil, 11, 14, nil, 4, nil, 9, 13, 5, 7, 3]
                  :rows [9, 12, 3, 13, nil, nil, 4, 8, 5, nil, 6, nil, 10]}]
      (is (= target (simulate current target)))))

  (testing "example 2"
    (let [current {:columns [8, 7, 15, nil, 11, 13, 2]
                   :rows [nil, 2, 6, 7, 8, 1, 5]}
          target {:columns [4, 5, nil, 2, 6, 1, 8, 7]
                  :rows [8, 6, 5, 7, 4, nil, 1]}]
      (is (= target (simulate current target)))))


  (testing "large sheet"
    (let [current (json/read-str (slurp "test/data/large_current.json") :key-fn keyword)
          target (json/read-str (slurp "test/data/large_target.json") :key-fn keyword)]
      (is (= target (simulate current target))))))
